src/java/org/apache/cassandra/service/reads/DataResolver.java

```diff
@@ -259,6 +259,7 @@ private PartitionIterator resolveWithReplicaFilteringProtection(E replicas, Repa
                                                                  i -> shortReadProtectedResponse(i, firstPhaseContext, null),
                                                                  UnaryOperator.identity());
 
+        // TODO: Filtering can cause short reads when a fetch size or LIMIT is low enough and we have to cull out possible matches from replicas.
```

Comment:
When replicas send tombstones those tombstones are not counted by the `LIMIT` truncation. It occurs to me that we could do something similar with rows using non-strict filtering. That is, we could count them for the limit if they satisfy the strict/AND filter, and don't count them if they only pass the non-strict/OR filter. 

That way, the replicas would send more rows than strictly required, potentially saving some SRP requests if the non-strict matches aren't indeed considered matches after reconciliation, which I think is the more common case.

Of course, using OR for the expressions in non-strict filtering can produce tons of results, so we should probably cap the number of non-strict matches that are included in the results out of the LIMIT count.